src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/auto_crafters/AbstractAutoCrafter.java

```diff
@@ -0,0 +1,432 @@
+import org.bukkit.block.Skull;
+import org.bukkit.entity.Player;
+import org.bukkit.inventory.Inventory;
+import org.bukkit.inventory.InventoryHolder;
+import org.bukkit.inventory.ItemStack;
+
+import io.github.thebusybiscuit.cscorelib2.data.PersistentDataAPI;
+import io.github.thebusybiscuit.cscorelib2.item.CustomItem;
+import io.github.thebusybiscuit.cscorelib2.protection.ProtectableAction;
+import io.github.thebusybiscuit.slimefun4.api.SlimefunAddon;
+import io.github.thebusybiscuit.slimefun4.api.items.ItemState;
+import io.github.thebusybiscuit.slimefun4.core.attributes.EnergyNetComponent;
+import io.github.thebusybiscuit.slimefun4.core.handlers.BlockUseHandler;
+import io.github.thebusybiscuit.slimefun4.core.networks.energy.EnergyNetComponentType;
+import io.github.thebusybiscuit.slimefun4.implementation.SlimefunPlugin;
+import io.github.thebusybiscuit.slimefun4.implementation.tasks.AsyncRecipeChoiceTask;
+import io.github.thebusybiscuit.slimefun4.utils.ChestMenuUtils;
+import io.papermc.lib.PaperLib;
+import me.mrCookieSlime.CSCoreLibPlugin.Configuration.Config;
+import me.mrCookieSlime.CSCoreLibPlugin.general.Inventory.ChestMenu;
+import me.mrCookieSlime.Slimefun.Lists.RecipeType;
+import me.mrCookieSlime.Slimefun.Objects.Category;
+import me.mrCookieSlime.Slimefun.Objects.SlimefunItem.SlimefunItem;
+import me.mrCookieSlime.Slimefun.Objects.SlimefunItem.abstractItems.AContainer;
+import me.mrCookieSlime.Slimefun.Objects.handlers.BlockTicker;
+import me.mrCookieSlime.Slimefun.api.SlimefunItemStack;
+
+/**
+ * This is the abstract super class for our auto crafters.
+ * 
+ * @author TheBusyBiscuit
+ * 
+ * @see VanillaAutoCrafter
+ * @see EnhancedAutoCrafter
+ *
+ */
+public abstract class AbstractAutoCrafter extends SlimefunItem implements EnergyNetComponent {
+
+    /**
+     * The amount of energy consumed per crafting operation.
+     */
+    private int energyConsumed = -1;
+
+    /**
+     * The amount of energy this machine can store.
+     */
+    private int energyCapacity = -1;
+
+    /**
+     * The {@link NamespacedKey} used to store recipe data.
+     */
+    protected final NamespacedKey recipeStorageKey;
+
+    // @formatter:off
+    protected final int[] background = {
+        0, 1, 2, 3, 4, 5, 6, 7, 8,
+        9, 10, 14, 15, 16, 17,
+        18, 19, 23, 25, 26,
+        27, 28, 32, 33, 34, 35,
+        36, 37, 38, 39, 40, 41, 42, 43, 44
+    };
+    // @formatter:on
+
+    @ParametersAreNonnullByDefault
+    public AbstractAutoCrafter(Category category, SlimefunItemStack item, RecipeType recipeType, ItemStack[] recipe) {
+        super(category, item, recipeType, recipe);
+
+        recipeStorageKey = new NamespacedKey(SlimefunPlugin.instance(), "recipe_key");
+
+        addItemHandler(onRightClick());
+        addItemHandler(new BlockTicker() {
+
+            @Override
+            public void tick(Block b, SlimefunItem item, Config data) {
+                AbstractAutoCrafter.this.tick(b, data);
```

Comment:
Im very confused about this line because you do `staticClass.this`
just wanted to point that out

Reply:
That is a pretty normal thing to do, it's done in almost every ticking Slimefun machine class too.